### PR TITLE
[Whisper] Fix word-level timestamps for audio < 30 seconds

### LINKED ISSUE #25605: Incorrect whisper word-level timestamps
### System Info

- `transformers` version: 4.32.0.dev0 (main)
- Platform: Linux-5.15.0-1041-azure-x86_64-with-glibc2.31
- Python version: 3.10.8
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.2
- Accelerate version: not installed
- Accelerate config: not found
- PyTorch version (GPU?): 2.0.1+cu117 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?: no

### Who can help?

@sanchit-gandhi

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Running
```python
from transformers import pipeline
pipe = pipeline("automatic-speech-recognition", "openai/whisper-base")
url = 'https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/japanese-audio.wav'
output = pipe(url, return_timestamps="word", chunk_length_s=30, generate_kwargs={'language': 'japanese'})
print(output)
```

Produces
```
{'text': '森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である', 'chunks': [{'text': '森', 'timestamp': (29.98, 29.98)}, {'text': '長', 'timestamp': (29.98, 29.98)}, {'text': 'の', 'timestamp': (29.98, 29.98)}, {'text': '美味', 'timestamp': (29.98, 29.98)}, {'text': 'しい', 'timestamp': (29.98, 29.98)}, {'text': '牛', 'timestamp': (29.98, 29.98)}, {'text': '乳', 'timestamp': (29.98, 29.98)}, {'text': 'は', 'timestamp': (29.98, 29.98)}, {'text': '濃', 'timestamp': (29.98, 29.98)}, {'text': 'い', 'timestamp': (29.98, 29.98)}, {'text': '青', 'timestamp': (29.98, 29.98)}, {'text': 'い', 'timestamp': (29.98, 29.98)}, {'text': 'ように', 'timestamp': (29.98, 29.98)}, {'text': '牛', 'timestamp': (29.98, 29.98)}, {'text': '乳', 'timestamp': (29.98, 29.98)}, {'text': 'ビ', 'timestamp': (29.98, 29.98)}, {'text': 'ーン', 'timestamp': (29.98, 29.98)}, {'text': 'を', 'timestamp': (29.98, 29.98)}, {'text': '足', 'timestamp': (29.98, 29.98)}, {'text': 'ら', 'timestamp': (29.98, 29.98)}, {'text': 'った', 'timestamp': (29.98, 29.98)}, {'text': '絶', 'timestamp': (29.98, 29.98)}, {'text': '対', 'timestamp': (29.98, 29.98)}, {'text': 'に', 'timestamp': (29.98, 29.98)}, {'text': 'の', 'timestamp': (29.98, 29.98)}, {'text': 'パ', 'timestamp': (29.98, 29.98)}, {'text': 'ック', 'timestamp': (29.98, 29.98)}, {'text': '牛', 'timestamp': (29.98, 29.98)}, {'text': '乳', 'timestamp': (29.98, 29.98)}, {'text': 'で', 'timestamp': (29.98, 29.98)}, {'text': 'ある', 'timestamp': (29.98, 29.98)}]}
```

### Expected behavior

Should not have all timestamps = 29.98. It should be noted that `29.98 = 30 - 0.02 = chunk_length_s - time_precision`, which I don't think is a coincidence.

Output should better-match the output from openai's whisper library:

 ```python
import whisper
model = whisper.load_model("base")
url = "https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/japanese-audio.wav"
result = model.transcribe(url, word_timestamps=True)
print(result)
```

```
{'text': '森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である', 'segments': [{'id': 0, 'seek': 0, 'start': 0.28, 'end': 6.06, 'text': '森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である', 'tokens': [50364, 16407, 106, 15353, 2972, 45511, 17121, 40003, 2930, 111, 3065, 26373, 225, 1764, 37462, 1764, 34483, 40003, 2930, 111, 39454, 44632, 5998, 37236, 5154, 10102, 6948, 114, 35252, 4108, 2972, 23268, 28551, 40003, 2930, 111, 2474, 24719, 50776], 'temperature': 0.0, 'avg_logprob': -0.5187992572784423, 'compression_ratio': 1.1142857142857143, 'no_speech_prob': 0.06956221163272858, 'words': [{'word': '森', 'start': 0.28, 'end': 0.64, 'probability': 0.6905971467494965}, {'word': '長', 'start': 0.64, 'end': 0.82, 'probability': 0.25675880908966064}, {'word': 'の', 'start': 0.82, 'end': 1.04, 'probability': 0.9857621192932129}, {'word': '美味', 'start': 1.04, 'end': 1.18, 'probability': 0.44574692845344543}, {'word': 'しい', 'start': 1.18, 'end': 1.48, 'probability': 0.9633633494377136}, {'word': '牛', 'start': 1.48, 'end': 1.68, 'probability': 0.9765644073486328}, {'word': '乳', 'start': 1.68, 'end': 1.94, 'probability': 0.8430313766002655}, {'word': 'は', 'start': 1.94, 'end': 2.12, 'probability': 0.9037646651268005}, {'word': '濃', 'start': 2.12, 'end': 2.34, 'probability': 0.7717265486717224}, {'word': 'い', 'start': 2.34, 'end': 2.46, 'probability': 0.9223815202713013}, {'word': '青', 'start': 2.46, 'end': 2.64, 'probability': 0.7740068435668945}, {'word': 'い', 'start': 2.64, 'end': 2.76, 'probability': 0.861002504825592}, {'word': 'ように', 'start': 2.76, 'end': 2.92, 'probability': 0.10338784009218216}, {'word': '牛', 'start': 2.92, 'end': 3.14, 'probability': 0.8816423416137695}, {'word': '乳', 'start': 3.14, 'end': 3.44, 'probability': 0.9983818531036377}, {'word': 'ビ', 'start': 3.44, 'end': 3.58, 'probability': 0.3014017343521118}, {'word': 'ーン', 'start': 3.58, 'end': 3.7, 'probability': 0.7804636359214783}, {'word': 'を', 'start': 3.7, 'end': 3.8, 'probability': 0.9925546050071716}, {'word': '足', 'start': 3.8, 'end': 3.98, 'probability': 0.26062318682670593}, {'word': 'ら', 'start': 3.98, 'end': 4.1, 'probability': 0.7511312365531921}, {'word': 'った', 'start': 4.1, 'end': 4.3, 'probability': 0.7001805901527405}, {'word': '絶', 'start': 4.3, 'end': 4.52, 'probability': 0.5361797697842121}, {'word': '対', 'start': 4.52, 'end': 4.68, 'probability': 0.27607205510139465}, {'word': 'に', 'start': 4.68, 'end': 4.78, 'probability': 0.40600043535232544}, {'word': 'の', 'start': 4.78, 'end': 4.94, 'probability': 0.8449609875679016}, {'word': 'パ', 'start': 4.94, 'end': 5.1, 'probability': 0.43969523906707764}, {'word': 'ック', 'start': 5.1, 'end': 5.22, 'probability': 0.916713297367096}, {'word': '牛', 'start': 5.22, 'end': 5.42, 'probability': 0.9680136442184448}, {'word': '乳', 'start': 5.42, 'end': 5.7, 'probability': 0.9995707273483276}, {'word': 'で', 'start': 5.7, 'end': 5.84, 'probability': 0.9711904525756836}, {'word': 'ある', 'start': 5.84, 'end': 6.06, 'probability': 0.9810820817947388}]}], 'language': 'ja'}
```

---

Fortunately, the output from the models are identical, so there's no issue with `generate`.

```
# transformers: 森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である
# openai: 森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である
```

### PR DESCRIPTION
**Pull request by xenova**
# What does this PR do?

In OpenAI's [original implementation for word-level timestamps](https://github.com/openai/whisper/blob/e8622f9afc4eba139bf796c210f5c01081000472/whisper/timing.py#L206), they crop the cross attentions before perform dynamic time warping (to only run the algorithm on valid audio; this prevents getting stuck when backtracking). The current transformers implementation misses this, so this PR fixes that.

Testing code:
```python
from transformers import pipeline
pipe = pipeline("automatic-speech-recognition", "openai/whisper-base")
url = 'https://huggingface.co/datasets/Xenova/transformers.js-docs/resolve/main/japanese-audio.wav'
output = pipe(url, return_timestamps="word", chunk_length_s=30, generate_kwargs={'language': 'japanese'})
print(output)
```

Fixed:
```
{
'text': '森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である',
'chunks': [
  {'text': '森', 'timestamp': (0.18, 0.64)},
  {'text': '長', 'timestamp': (0.64, 0.82)},
  {'text': 'の', 'timestamp': (0.82, 1.04)},
  {'text': '美味', 'timestamp': (1.04, 1.2)},
  {'text': 'しい', 'timestamp': (1.2, 1.46)},
  {'text': '牛', 'timestamp': (1.46, 1.68)},
  {'text': '乳', 'timestamp': (1.68, 1.92)},
  {'text': 'は', 'timestamp': (1.92, 2.14)},
  {'text': '濃', 'timestamp': (2.14, 2.32)},
  {'text': 'い', 'timestamp': (2.32, 2.44)},
  {'text': '青', 'timestamp': (2.44, 2.64)},
  {'text': 'い', 'timestamp': (2.64, 2.76)},
  {'text': 'ように', 'timestamp': (2.76, 2.92)},
  {'text': '牛', 'timestamp': (2.92, 3.16)},
  {'text': '乳', 'timestamp': (3.16, 3.36)},
  {'text': 'ビ', 'timestamp': (3.36, 3.58)},
  {'text': 'ーン', 'timestamp': (3.58, 3.66)},
  {'text': 'を', 'timestamp': (3.66, 3.82)},
  {'text': '足', 'timestamp': (3.82, 4.0)},
  {'text': 'ら', 'timestamp': (4.0, 4.12)},
  {'text': 'った', 'timestamp': (4.12, 4.3)},
  {'text': '絶', 'timestamp': (4.3, 4.52)},
  {'text': '対', 'timestamp': (4.52, 4.68)},
  {'text': 'に', 'timestamp': (4.68, 4.78)},
  {'text': 'の', 'timestamp': (4.78, 4.94)},
  {'text': 'パ', 'timestamp': (4.94, 5.1)},
  {'text': 'ック', 'timestamp': (5.1, 5.2)},
  {'text': '牛', 'timestamp': (5.2, 5.44)},
  {'text': '乳', 'timestamp': (5.44, 5.64)},
  {'text': 'で', 'timestamp': (5.64, 5.84)},
  {'text': 'ある', 'timestamp': (5.84, 6.04)}
]
}
```

Previous (broken):
```
{
'text': '森長の美味しい牛乳は濃い青いように牛乳ビーンを足らった絶対にのパック牛乳である',
'chunks': [
  {'text': '森', 'timestamp': (29.98, 29.98)},
  {'text': '長', 'timestamp': (29.98, 29.98)},
  {'text': 'の', 'timestamp': (29.98, 29.98)},
  {'text': '美味', 'timestamp': (29.98, 29.98)},
  {'text': 'しい', 'timestamp': (29.98, 29.98)},
  {'text': '牛', 'timestamp': (29.98, 29.98)},
  {'text': '乳', 'timestamp': (29.98, 29.98)}, 
  'text': 'は', 'timestamp': (29.98, 29.98)},
  {'text': '濃', 'timestamp': (29.98, 29.98)},
  {'text': 'い', 'timestamp': (29.98, 29.98)},
  {'text': '青', 'timestamp': (29.98, 29.98)},
  {'text': 'い', 'timestamp': (29.98, 29.98)},
  {'text': 'ように', 'timestamp': (29.98, 29.98)},
  {'text': '牛', 'timestamp': (29.98, 29.98)},
  {'text': '乳', 'timestamp': (29.98, 29.98)},
  {'text': 'ビ', 'timestamp': (29.98, 29.98)},
  {'text': 'ーン', 'timestamp': (29.98, 29.98)},
  {'text': 'を', 'timestamp': (29.98, 29.98)},
  {'text': '足', 'timestamp': (29.98, 29.98)},
  {'text': 'ら', 'timestamp': (29.98, 29.98)},
  {'text': 'った', 'timestamp': (29.98, 29.98)},
  {'text': '絶', 'timestamp': (29.98, 29.98)},
  {'text': '対', 'timestamp': (29.98, 29.98)},
  {'text': 'に', 'timestamp': (29.98, 29.98)},
  {'text': 'の', 'timestamp': (29.98, 29.98)},
  {'text': 'パ', 'timestamp': (29.98, 29.98)},
  {'text': 'ック', 'timestamp': (29.98, 29.98)},
  {'text': '牛', 'timestamp': (29.98, 29.98)},
  {'text': '乳', 'timestamp': (29.98, 29.98)},
  {'text': 'で', 'timestamp': (29.98, 29.98)},
  {'text': 'ある', 'timestamp': (29.98, 29.98)}
]
}
```

Fixes #25605  (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi @ArthurZucker

### CODE REVIEW

**Comment by xenova** [comment]
Looks like there were some failing unit tests, but they were actually wrong 😅 

### Original unit test:
```python
{
    "text": " Conquered returned to its place amidst the tents.",
    "chunks": [
        {"text": " Conquered", "timestamp": (29.78, 29.9)},
        {"text": " returned", "timestamp": (29.9, 29.9)},
        {"text": " to", "timestamp": (29.9, 29.9)},
        {"text": " its", "timestamp": (29.9, 29.9)},
        {"text": " place", "timestamp": (29.9, 29.9)},
        {"text": " amidst", "timestamp": (29.9, 29.9)},
        {"text": " the", "timestamp": (29.9, 29.9)},
        {"text": " tents.", "timestamp": (29.9, 29.9)}
    ]
}
```

### New (fixed) unit test:
```python
{
    "text": " Conquered returned to its place amidst the tents.",
    "chunks": [
        {"text": " Conquered", "timestamp": (0.54, 1.2)},
        {"text": " returned", "timestamp": (1.2, 1.62)},
        {"text": " to", "timestamp": (1.62, 1.88)},
        {"text": " its", "timestamp": (1.88, 2.02)},
        {"text": " place", "timestamp": (2.02, 2.3)},
        {"text": " amidst", "timestamp": (2.3, 2.84)},
        {"text": " the", "timestamp": (2.84, 2.98)},
        {"text": " tents.", "timestamp": (2.98, 3.5)},
    ],
},
```

File: src/transformers/models/whisper/modeling_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/modeling_whisper.py:1609]:
```suggestion
            num_frames (`float`, *optional*):
                The number of audio frames available in this chunk. This is only used when generating word-level timestamps.
```
I think this should go in the generation config WDYT? (Since you could generate outside the pipeline, with a predifined number of frames.

File: src/transformers/models/whisper/modeling_whisper.py
Comment (ArthurZucker) [inline src/transformers/models/whisper/modeling_whisper.py:1772]:
```suggestion
                outputs, generation_config.alignment_heads, num_frames=generation_config.num_frames
```

**Review by ArthurZucker — COMMENTED** [review]
Very nice catch! Will leave @sanchit-gandhi have a look!

File: src/transformers/models/whisper/modeling_whisper.py
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/modeling_whisper.py:1609]: Yep this can go in the generation config! As we have with `generation_config.alignment_heads` for example

File: src/transformers/models/whisper/modeling_whisper.py
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_whisper.py:1823]: Nice catch!

File: src/transformers/models/whisper/modeling_whisper.py
Comment (sanchit-gandhi) [reply src/transformers/models/whisper/modeling_whisper.py:1609]: Note that the nice thing here is that even if we use the generation config to store this property, it will still be updated by any kwargs we pass to the `.generate` method.

E.g. if we set `generation_config.num_frames=None`, but do:
```python
model.generate(input_features, num_frames=num_frames)
```

Then the generation config will be updated with the value of `num_frames` we pass to `.generate`

=> this way, `num_frames` still behaves like a kwarg, but is actually just updated in the `generation_config`

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (sanchit-gandhi) [inline src/transformers/pipelines/automatic_speech_recognition.py:533]: If `stride` is None (i.e. we don't use the chunking algorithm) then `num_frames` is also `None`? We can set the default for `num_frames` in the `generation_config` to the value we'd use **without** any chunking

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py
Comment (sanchit-gandhi) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:327]: We should have known!

**Review by sanchit-gandhi — COMMENTED** [review]
Looks really nice already @xenova! Thanks for the fix! Echo'ing @ArthurZucker's suggestion about using the generation config, but otherwise looks top 👌

File: src/transformers/models/whisper/modeling_whisper.py
Comment (xenova) [reply src/transformers/models/whisper/modeling_whisper.py]: Will add to generation config! 👍

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (xenova) [reply src/transformers/pipelines/automatic_speech_recognition.py:533]: `num_frames=None` should emulate the default behaviour: i.e., crop by 3000 frames (which is the same as just not cropping).

Comment (xenova) [reply src/transformers/pipelines/automatic_speech_recognition.py:533]: In particular,

> We can set the default for num_frames in the generation_config

wouldn't matter since not cropping is the same as using the default crop of 3000.

Comment (sanchit-gandhi) [reply src/transformers/pipelines/automatic_speech_recognition.py:533]: Ah I see - thanks for the clarification! Good for me in that case

**Review by sanchit-gandhi — APPROVED** [review]
Lovely thanks @xenova! LGTM! Feel free to merge @ArthurZucker if you're happy with the PR

**Comment by sanchit-gandhi** [comment]
Gently pinging @ArthurZucker for the final 👍 before merge - thank you again for the PR @xenova!

**Comment by ArthurZucker** [comment]
Sorry for the miss reviewing now!

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (ArthurZucker) [inline src/transformers/pipelines/automatic_speech_recognition.py:531]: Only comment is that we end up checking for the stride thrice:
-[ here ](https://github.com/huggingface/transformers/pull/25607/files#diff-ccc8e98fcaa81fdf6317a652438a309bcede0bbe336774288c2fcf91d9f11082R566)

- and [here](https://github.com/huggingface/transformers/pull/25607/files#diff-ccc8e98fcaa81fdf6317a652438a309bcede0bbe336774288c2fcf91d9f11082R571)

so let's just do it once and forall 😉

**Review by ArthurZucker — APPROVED** [review]
Sorry for the wait! Looks super good to me! 
Left a very small nit 😉 
Thanks for fixing

**Comment by aramfaghfouri** [comment]
Thank you for your contribution @xenova!
Is this PR approved and merged in the latest version? I just installed Transformers and I am still getting the old results:
{'text': ' Okay, you ready?', 'chunks': [{'text': ' Okay,', 'timestamp': (29.98, 29.98)}, {'text': ' you', 'timestamp': (29.98, 29.98)}, {'text': ' ready?', 'timestamp': (29.98, 29.98)}]}

Thanks!

**Comment by xenova** [comment]
Hi there - It's not yet merged, but will hopefully be soon!

**Comment by aramfaghfouri** [comment]
Do you happen to know when the next version will be out?

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (sanchit-gandhi) [inline src/transformers/pipelines/automatic_speech_recognition.py:546]: We weren't doing anything special here, just pop'ing the attention mask from the inputs

=> I've moved the attention mask to the top of this method so that we only compute it once

**Review by sanchit-gandhi — COMMENTED** [review]
Quickly pushed the last requested changes! Will merge when the CI is green!

File: src/transformers/pipelines/automatic_speech_recognition.py
Comment (sanchit-gandhi) [reply src/transformers/pipelines/automatic_speech_recognition.py]: Resolved in https://github.com/huggingface/transformers/pull/25607/commits/0694dbe3e597daf973e3f1545b9fcc2e0da4afdf

**Comment by aramfaghfouri** [comment]
Hi @xenova,

I just upgraded to the latest version of Transformers (4.33.1) and tried the following. It seems that the word timestamps are still incorrect.
What am I doing wrong?

Thanks!

###########
pipe = pipeline(
  "automatic-speech-recognition",
  model="openai/whisper-large",
  chunk_length_s=4,
  device=device,
)
vx0 = pipe(x0, batch_size=4, return_timestamps="word")

vx0['chunks']:
{'text': ' Okay,', 'timestamp': (29.98, 29.98)},
 {'text': ' you', 'timestamp': (29.98, 29.98)},
 {'text': ' ready?', 'timestamp': (29.98, 29.98)},
 {'text': ' So', 'timestamp': (32.65, 32.65)},
 {'text': ' now', 'timestamp': (32.65, 32.65)},
 {'text': ' it', 'timestamp': (35.31, 35.31)},
 {'text': ' is', 'timestamp': (35.31, 35.31)},
 {'text': " it's", 'timestamp': (35.31, 35.31)},
 {'text': ' a', 'timestamp': (35.31, 35.31)},

**Comment by xenova** [comment]
This is because a full release hasn't come out yet. To fix it, you can install from source (see [docs](https://huggingface.co/docs/transformers/installation#install-from-source)):
```
pip install --upgrade git+https://github.com/huggingface/transformers
```

**Comment by Starlento** [comment]
Hi guys, I just tried whisper v3 and find that your updated code is gone in the current main branch.
And it gives me 29.98... There are multiple commits after this merge, can someone check what is going on?

**Pull request merged**
